andersoncd/weighted_lasso.py

```diff
@@ -0,0 +1,394 @@
+                R = y - X @ w
+            else:
+                w = np.zeros(n_features, dtype=X.dtype)
+                R = y.copy()
+
+        sol = celer_primal(
+            X, y, alpha, w, R, norms_X_col, weights,
+            max_iter=max_iter, max_epochs=max_epochs, p0=p0, tol=tol,
+            verbose=verbose)
+
+        coefs[:, t] = w.copy()
+        kkt_maxs[t] = sol[-1]
+        # TODO sol[0], sol[1], sol[2][-1]
+        if return_n_iter:
+            n_iters[t] = len(sol[1])
+
+    results = alphas, coefs, kkt_maxs
+    if return_n_iter:
+        results += (n_iters,)
+
+    return results
+
+
+# def _kkt_violation(XtR, weights, alpha):
+#     for j in range(n_features):
+#         if weights[j] > 0:
+#             kkt[j] = max(0, np.abs(XtR[j]) / n_samples - alpha * weights[j]))
+#         else:
+#                 kkt_max=max(kkt_max, np.abs(XtR[j]))
+
+
+def celer_primal(
+        X, y, alpha, w, R, norms_X_col, weights, max_iter=50,
+        max_epochs=50_000, p0=10, tol=1e-4, use_acc=False, K=5, verbose=0):
+    n_samples, n_features = X.shape
+    pen = weights > 0
+    unpen = ~pen
+    n_unpen = unpen.sum()
+    print(n_unpen)
+    p0 = max(p0, n_unpen)
+    obj_out = []
+    lc = norms_X_col ** 2
+
+    if use_acc:
+        last_K_w = np.zeros([K + 1, n_features])
+        U = np.zeros([K, n_features])
+
+    for t in range(max_iter):
+        kkt = _kkt_violation(w, X, R, weights, alpha, np.arange(n_features))
+        kkt_max = np.max(kkt)
+        if verbose:
+            print(f"KKT max violation: {kkt_max:.2e}")
+        if kkt_max <= tol:
+            break
+        # 1) select features
+        ws_size = max(p0 + n_unpen,
+                      min(2 * (w != 0).sum() - n_unpen, n_features))
+        kkt[unpen] = np.inf  # always include unpenalized features
+        ws = np.argsort(kkt)[-ws_size:]
+
+        if verbose:
+            print(f'Iteration {t}, {ws_size} feats in subpb.')
+
+        # 2) do iterations on smaller problem
+        for epoch in range(max_epochs):
+            _cd_wlasso(X, w, R, alpha, weights, lc, ws)
+
+            if use_acc:
+                last_K_w[epoch % (K + 1)] = w
+
+                if epoch % (K + 1) == K:
+                    for k in range(K):
+                        U[k] = last_K_w[k + 1] - last_K_w[k]
+                    C = np.dot(U, U.T)
+
+                    try:
+                        z = np.linalg.solve(C, np.ones(K))
+                        c = z / z.sum()
+                        w_acc = np.sum(last_K_w[:-1] * c[:, None], axis=0)
+                        p_obj = primal_wlasso(R, w, alpha, weights)
+                        R_acc = y - X @ w_acc
+                        p_obj_acc = primal_wlasso(R_acc, w_acc, alpha, weights)
+                        if p_obj_acc < p_obj:
+                            w = w_acc
+                            R = R_acc
+                    except np.linalg.LinAlgError:
+                        if verbose:
```

Comment:
I have verbose which is an int, here verbose >= 2 means verbose inner solver, so `if max(verbose - 1, 0)` rather than verbose